core/jobs/batch_jobs/audit_rule_validation_jobs.py

```diff
@@ -0,0 +1,2194 @@
+        for state_name, state in states_dict.items():
+            if state.interaction.id != 'NumericInput':
+                continue
+            answer_groups = state.interaction.answer_groups
+            invalid_ans_group_count = 0
+            valid_dest_ans_group_count = 0
+            invalid_ans_groups = []
+            rule_spec_till_now: List[state_domain.RuleSpecDict] = []
+            ranges: List[RangeVariable] = []
+            for ans_group_idx, answer_group in enumerate(answer_groups):
+                can_delete_ans_group, ranges = (
+                    ExpAuditRuleChecksJob.
+                    _filter_invalid_numeric_ans_group(
+                        answer_group, rule_spec_till_now, ans_group_idx,
+                        ranges
+                    )
+                )
+
+                for rule_spec in answer_group.rule_specs:
+                    rule_spec_till_now.append(rule_spec.to_dict())
+
+                if answer_group.outcome.dest != state_name:
+                    valid_dest_ans_group_count += 1
+                    if can_delete_ans_group:
+                        invalid_ans_group_count += 1
+                        invalid_ans_groups.append(ans_group_idx)
+
+            if (
+                invalid_ans_group_count == valid_dest_ans_group_count and
+                invalid_ans_group_count != 0 and
+                valid_dest_ans_group_count != 0
+            ):
+                invalid_states.append(
+                    {
+                        'state_name': state_name,
+                        'ans_group_idx': invalid_ans_groups
+                    }
+                )
+
+        return invalid_states
+
+    @staticmethod
+    def _should_check_range_criteria(
+        earlier_rule: state_domain.RuleSpec,
+        later_rule: state_domain.RuleSpec
+    ) -> bool:
+        """Checks the range criteria between two rules by comparing their
+        rule type
+
+        Args:
+            earlier_rule: state_domain.RuleSpec. Previous rule.
+            later_rule: state_domain.RuleSpec. Current rule.
+
+        Returns:
+            bool. Returns True if the rules passes the range criteria check.
+        """
+        if earlier_rule.rule_type != 'IsExactlyEqualTo':
+            return True
+        return later_rule.rule_type not in (
+            'IsExactlyEqualTo', 'IsEquivalentTo',
+            'IsEquivalentToAndInSimplestForm'
+        )
+
+    @staticmethod
+    def _get_rule_value_f(
+        rule_spec: state_domain.RuleSpec
+    ) -> Any:
+        """Return rule values from the rule_spec
+
+        Args:
+            rule_spec: state_domain.RuleSpec. Rule spec of an answer group.
+
+        Returns:
+            rule_value_f: Any. The value of the rule spec.
+        """
+        rule_value_f = rule_spec.inputs['f']
+        if rule_value_f['wholeNumber'] == 0:
+            rule_value_f = float(
+                rule_value_f['numerator'] / rule_value_f['denominator']
+            )
```

Comment:
You don't need to convert to `float` in division in Python.

Reply:
Yes but the values can be in `int` and for the fraction interaction to keep the track of the `ranges` of the rules, I'm converting them to float and then keeping the track of them

Reply:
Okay. In Python, the integer `2` is equal to the float `2.0`. So you don't need `float`, but it's okay 🤷